### PR TITLE
Add Grad-CAM VGG16 ImageNet project

### LINKED ISSUE #1539: Grad-CAM Implementation Using VGG16 on ImageNet
Currently, the repository contains many deep learning and image classification projects, but there is limited focus on Explainable AI (XAI) techniques that help users understand how CNN models make predictions. Model interpretability is an important aspect of modern AI systems, especially in research and real-world applications.

I would like to add a Grad-CAM (Gradient-weighted Class Activation Mapping) project using a pre-trained VGG16 model on the ILSVRC (ImageNet) dataset. The project will generate heatmaps highlighting the important regions of an image that influence the model’s prediction.

The implementation follows the paper:

Selvaraju et al.,
"Grad-CAM: Visual Explanations from Deep Networks via Gradient-based Localization", ICCV 2017.

The project will include:

Grad-CAM implementation notebook
Heatmap visualization
Overlay of heatmaps on original images
Documentation explaining the workflow

Other explainability methods such as CAM and Grad-CAM++ were considered. However, Grad-CAM was chosen because it is simpler to understand, widely used, and works effectively with pre-trained CNN architectures like VGG16.

Technologies used: Python, TensorFlow/Keras, OpenCV, Matplotlib
Implemented in Google Colab
Uses sample ImageNet images for demonstration purposes
This project can help beginners understand CNN interpretability and Explainable AI concepts
I would like to work on this issue and submit a Pull Request.

### PR DESCRIPTION
**Pull request by Aditri-web**
## Add Grad-CAM VGG16 ImageNet Project

This PR adds the `Grad_CAM_VGG16_ImageNet` folder containing a complete implementation of Gradient-weighted Class Activation Mapping (Grad-CAM) using a pre-trained VGG16 model on ImageNet.

### Files Added
- `GRAD_CAM.ipynb` - Jupyter notebook demonstrating Grad-CAM visualization
- `grad_cam.py` - Python script for generating Grad-CAM heatmaps
- `README.md` - Project overview, features, and references
- `requirements.txt` - Dependencies: tensorflow, opencv-python, matplotlib, numpy

### About
Grad-CAM (Gradient-weighted Class Activation Mapping) is a technique that produces visual explanations for CNN decisions by using the gradients of target class flowing into the final convolutional layer.

**Reference:** Selvaraju et al., "Grad-CAM: Visual Explanations from Deep Networks via Gradient-based Localization", ICCV 2017

### CODE REVIEW

File: Grad_CAM_VGG16_ImageNet/GRAD_CAM.ipynb
Comment (Sibam-Paul) [inline Grad_CAM_VGG16_ImageNet/GRAD_CAM.ipynb:1]: Image

ur secret is out in the open 
create a .env  for secrets example

just add a button in the notebook that says open in collab .. coz ur using google collab

also explain the notebook use markdown to explain the code and some txt

File: Grad_CAM_VGG16_ImageNet/grad_cam.py [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+
+    return cam, target_class
+
+
+# Test with sample images from the downloaded ImageNet mini dataset
+# Using local paths to avoid rate-limiting from external URLs
+test_image_paths = [
+    "/content/data/imagenet-mini/train/n02099712/n02099712_1000.JPEG", # Labrador retriever
+    "/content/data/imagenet-mini/train/n02123597/n02123597_1000.JPEG", # Siamese cat
+    "/content/data/imagenet-mini/train/n02504458/n02504458_1000.JPEG", # African elephant
```
Comment (Sibam-Paul) [inline Grad_CAM_VGG16_ImageNet/grad_cam.py:340]: for the test images add a pdf explaining td images @Aditri-web

File: Grad_CAM_VGG16_ImageNet/README.md
Comment (Sibam-Paul) [inline Grad_CAM_VGG16_ImageNet/README.md:1]: add steps how to run the files 
and how to use the py file and the notebook

Comment (Sibam-Paul) [reply Grad_CAM_VGG16_ImageNet/README.md:1]: for the beginners to use ur project @Aditri-web

**Review by Sibam-Paul — CHANGES_REQUESTED** [review]

**Comment by Aditri-web** [comment]
Hey @Sibam-Paul! I've addressed all the review comments — here's a quick summary of what was fixed:

Secret removed — hardcoded Kaggle token is gone; added .env.example + .gitignore so credentials are loaded safely via Colab Secrets or a local .env file
Open in Colab button added to the README
README fully rewritten with beginner-friendly step-by-step run instructions (both Colab and local), a code walkthrough, and a secrets setup guide
test_images_explanation.md added — explains all 3 test image classes (Labrador, Siamese cat, African elephant), what Grad-CAM highlights, and how to read the heatmap
CI/CD fixes — all deprecated GitHub Actions upgraded, permissions block added to fix the code scanning errors
Could you please re-review when you get a chance? Happy to make any further changes!

File: .github/workflows/build.yml
Comment (Sibam-Paul) [inline .github/workflows/build.yml:1]: ok

File: .github/workflows/shiftleft-analysis.yml
Comment (Sibam-Paul) [inline .github/workflows/shiftleft-analysis.yml:1]: ok

**Review by Sibam-Paul — APPROVED** [review]
was an error but the notebook cells has been run
LGTM

**Comment by Aditri-web** [comment]
Hey @Sibam-Paul! One more fix pushed — addressed the notebook cell outputs issue 🧹

GRAD_CAM.ipynb was committed with all executed outputs embedded (images, logs, print statements)
Cleared all cell outputs using jupyter nbconvert --ClearOutputPreprocessor before committing
File size dropped from 3.2 MB → 26 KB
Going forward, I've also added nbstripout as a recommendation in the README so outputs are never accidentally committed again.

All review points are now resolved

**Comment by Sibam-Paul** [comment]
we need output bro @Aditri-web 
remove the nbstripout

**Comment by Sibam-Paul** [comment]
see the pinned issues for guidelines @Aditri-web

File: Grad_CAM_VGG16_ImageNet/GRAD_CAM.ipynb
Comment (Sibam-Paul) [reply Grad_CAM_VGG16_ImageNet/GRAD_CAM.ipynb:1]: all cells must be run

**Review by Sibam-Paul — CHANGES_REQUESTED** [review]

**Comment by Sibam-Paul** [comment]
@Aditri-web

**Comment by Aditri-web** [comment]
Hey @Sibam-Paul! I've restored the notebook outputs as requested 📊
- Restored `GRAD_CAM.ipynb` to its fully run state so all visualization outputs, images, and logs are now embedded in the file.
- Replaced the hardcoded Kaggle API Token in the notebook credential setup cell with a secure fallback that dynamically loads it from an uploaded `kaggle.json` file or Colab Secrets.
Please re-review and approve when you have a moment. Ready to merge!

**Review by Sibam-Paul — APPROVED** [review]
ok
